Source/AssetRipper.Processing/AnimatorControllers/AnimatorStateContext.cs

```diff
@@ -0,0 +1,365 @@
+		{
+			IChildMotion childMotion = tree.Childs.AddNew();
+			IBlendTreeConstant treeConstant = state.GetBlendTree();
+			IBlendTreeNodeConstant node = treeConstant.NodeArray[nodeIndex].Data;
+			int childNodeIndex = (int)node.ChildIndices[childIndex];
+			IMotion? motion = CreateMotion(file, controller, state, childNodeIndex);
+			childMotion.Motion.SetAsset(tree.Collection, motion);
+
+			IBlendTreeNodeConstant childNode = treeConstant.NodeArray[childNodeIndex].Data;
+			if (childNode.IsBlendTree())
+			{
+				// BlendTree ChildMotions are not allowed to use TimeScale or Mirror
+				// https://github.com/Unity-Technologies/UnityCsReference/blob/4e215c07ca8e9a32a589043202fd919bdfc0a26d/Editor/Mono/Inspector/BlendTreeInspector.cs#L1469
+				// https://github.com/Unity-Technologies/UnityCsReference/blob/4e215c07ca8e9a32a589043202fd919bdfc0a26d/Editor/Mono/Inspector/BlendTreeInspector.cs#L1488
+				childMotion.TimeScale = 1;
+				childMotion.Mirror = false;
+			}
+			else
+			{
+				childMotion.TimeScale = 1 / childNode.Duration;
+				childMotion.Mirror = childNode.Mirror;
+			}
+			childMotion.CycleOffset = childNode.CycleOffset;
+
+			childMotion.Threshold = node.GetThreshold(childIndex);
+			childMotion.Position?.CopyValues(node.GetPosition(childIndex));
+			if (node.TryGetDirectBlendParameter(childIndex, out uint directID))
+			{
+				childMotion.DirectBlendParameter = controller.TOS[directID];
+			}
+
+			return childMotion;
+		}
+
+		private static string MakeStateMachinePath(AssetDictionary<uint, Utf8String> TOS, uint statePathID, string stateName)
+		{
+			string fullPath = TOS[statePathID];
+			string stateMachinePath = fullPath[..(fullPath.Length - stateName.Length - 1)];
+			return stateMachinePath;
+		}
+
+		/// <summary>
+		/// Create all AnimatorStates, and find names for StateMachines
+		/// </summary>
+		public AnimatorStateContext(ProcessedAssetCollection virtualFile, IAnimatorController controller, IStateMachineConstant stateMachineConstant, int layerIndex)
+		{
+			if (!controller.TOS.ContainsKey(0))
+			{
+				controller.TOS[0] = Utf8String.Empty;
+			}
+
+			DefaultStateIdx = stateMachineConstant.DefaultState != uint.MaxValue ? (int)stateMachineConstant.DefaultState : 0;
+			StateCount = stateMachineConstant.StateConstantArray.Count;
+
+			stateMachinePathNamesAndIDs = new();
+			if (!HasStates())
+			{
+				stateConstants = [];
+				states = [];
+				stateIdxsToStateMachinePathIDs = [];
+				return;
+			}
+			stateConstants = new IStateConstant[StateCount];
+			states = new IAnimatorState[StateCount];
+			stateIdxsToStateMachinePathIDs = new uint[StateCount];
+
+			for (int i = 0; i < StateCount; i++)
+			{
+				IStateConstant stateConstant = stateMachineConstant.StateConstantArray[i].Data;
+				IAnimatorState state = CreateAnimatorState(virtualFile, controller, controller.TOS, layerIndex, stateConstant);
```

Comment:
Asset creation should not be happening in the constructor of a context class because it's unintuitive.